File: ruby/knapsack.rb

class Packer

  def initialize(values, weights)
    @values = values
    @weights = weights

    raise 'dumb' if @values.size != @weights.size

    @memos = {}
  end

  def knapsack(capacity)
    pack_bag(@values.size - 1, capacity)
  end

  private

  def pack_bag(index, capacity)
    return 0 if index < 0 || capacity == 0

    if value_memoized?(index, capacity)
      puts "Value for index #{index} and capacity #{capacity } is stored"
      return @memos[index][capacity]
    else
      puts "Calling for index #{index}, Capacity left #{capacity}"
    end

    if @weights[index] > capacity
      result = pack_bag(index - 1, capacity)
    else
      added_value = pack_bag(index - 1, capacity - @weights[index]) + @values[index]
      not_added_value = pack_bag(index - 1, capacity)
      result = [added_value, not_added_value].max
    end

    memoize_value(index, capacity, result)

    result
  end

  def memoize_value(index, capacity, result)
    @memos[index] ||= {}
    @memos[index][capacity] = result
  end

  def value_memoized?(index, capacity)
    @memos[index] && @memos[index][capacity]
  end

end

p = Packer.new([1,2,3,2,5,6,3,2,1,2,3,1,2], [1,2,3,4,6,3,4,2,4,2,4,2,1])
puts p.knapsack(10)

# W V
# 1 3
# 2 2
# 3 2

File: js_output/knapsack.js
/**
 * Category: DYNAMIC_PROGRAMMING
 * Tags: recursion maps algorithms
 *
 * Given a set of objects, each of which has a weight and a monetary value,
 * put as many objects into a knapsack with a weight limit where the total value
 * of all the objects is the highest possible.
 */
import { assert } from "console";
class Packer {
    constructor(v, w) {
        this.values = v;
        this.weights = w;
        if (this.values.length !== this.weights.length) {
            throw "Values and Weights must be the same length";
        }
        this.memos = new Map();
    }
    pack(capacity) {
        return this.packHelper(this.values.length - 1, capacity);
    }
    packHelper(index, capacity) {
        if (index < 0 || capacity === 0)
            return 0;
        if (this.alreadyCalculated(index, capacity)) {
            return this.memos.get(index).get(capacity);
        }
        let result;
        if (this.weights[index] > capacity) {
            result = this.packHelper(index - 1, capacity);
        }
        else {
            const addedValue = this.packHelper(index - 1, capacity - this.weights[index]) +
                this.values[index];
            const notAddedValue = this.packHelper(index - 1, capacity);
            result = Math.max(addedValue, notAddedValue);
        }
        this.memoizeResult(index, capacity, result);
        return result;
    }
    alreadyCalculated(index, capacity) {
        return this.memos.has(index) && this.memos.get(index).has(capacity);
    }
    memoizeResult(index, capacity, result) {
        if (!this.memos.has(index)) {
            this.memos.set(index, new Map());
        }
        this.memos.get(index).set(capacity, result);
    }
}
const p = new Packer([1, 2, 12, 10], [1, 2, 6, 7]);
assert(p.pack(12) === 15);
const p2 = new Packer([1, 2, 3, 2, 5, 6, 3, 2, 1, 2, 3, 1, 2], [1, 2, 3, 4, 6, 3, 4, 2, 4, 2, 4, 2, 1]);
assert(p2.pack(10) === 14);
console.log("All assertions passed");
